File: www/warden/Warden.php
<?php
///////////////////////////////////////////
// author     : Ben Kietzman
// begin      : 2021-04-15
// copyright  : Ben Kietzman
// email      : ben@kietzman.org
///////////////////////////////////////////
/*! \file Warden.php
* \brief Warden Class
*/
class Warden
{
  // {{{ variables
  protected $m_strApplication;
  protected $m_strError;
  protected $m_strUnix;
  // }}}
  // {{{ __construct()
  public function __construct($strApplication, $strUnix)
  {
    $this->m_strUnix = $strUnix;
    if ($strApplication != '')
    {
      $this->m_strApplication = $strApplication;
    }
    else
    {
      $this->m_strError = 'Please provide the Application.';
    }
  }
  // }}}
  // {{{ __destruct()
  public function __destruct()
  {
  }
  // }}}
  // {{{ getError()
  public function getError()
  {
    return $this->m_strError;
  }
  // }}}
  // {{{ authn()
  public function authn(&$data)
  {
    return $this->request('authn', $data);
  }
  // }}}
  // {{{ authz()
  public function authz(&$data)
  {
    return $this->request('authz', $data);
  }
  // }}}
  // {{{ bridge
  // {{{ bridge()
  public function bridge(&$data)
  {
    return $this->request('bridge', $data);
  }
  // }}}
  // {{{ bridgeAuthn()
  public function bridgeAuthn($strUser, $strPassword)
  {
    $bResult = false;

    $data = [];
    $bResult = $this->bridgeAuthz($strUser, $strPassword, $data);
    unset($data);

    return $bResult;
  }
  // }}}
  // {{{ bridgeAuthz()
  public function bridgeAuthz($strUser, $strPassword, &$data)
  {
    $data['User'] = $strUser;
    $data['Password'] = $strPassword;

    return $this->bridge($data);
  }
  // }}}
  // }}}
  // {{{ central
  // {{{ central()
  public function central(&$data)
  {
    return $this->request('central', $data);
  }
  // }}}
  // {{{ centralUser()
  public function centralUser($strUser, &$data)
  {
    $data['User'] = $strUser;

    return $this->central($data);
  }
  // }}}
  // }}}
  // {{{ password
  // {{{ password()
  public function password(&$data)
  {
    return $this->request('password', $data);
  }
  // }}}
  // {{{ passwordApplication()
  public function passwordApplication($strApplication, $strUser, $strPassword, $strType)
  {
    $bResult = false;

    $data = [];
    $data['Application'] = $strApplication;
    $data['User'] = $strUser;
    $data['Password'] = $strPassword;
    if ($strType != '')
    {
      $data['Type'] = $strType;
    }
    $bResult = $this->password($data);
    unset($data);

    return $bResult;
  }
  // }}}
  // {{{ passwordUser()
  public function passwordUser($strUser, $strPassword)
  {
    $bResult = false;

    $data = [];
    $data['User'] = $strUser;
    $data['Password'] = $strPassword;
    $bResult = $this->password($data);
    unset($data);

    return $bResult;
  }
  // }}}
  // }}}
  // {{{ request
  // {{{ request()
  public function request($strModule, &$data)
  {
    $bResult = false;

    $request = $data;
    $data = [];
    $request['Module'] = $strModule;
    $response = null;
    if ($this->requestFull($request, $response))
    {
      $bResult = true;
      if (is_array($response) && isset($response['Data']))
      {
        $data = $response['Data'];
      }
    }
    unset($request);
    unset($response);

    return $bResult;
  }
  // }}}
  // {{{ requestFull()
  public function requestFull($request, &$response)
  {
    $bResult = false;
    $handle = null;
    $nErrorNo = null;
    $strError = null;

    $response = [];
    if (($handle = fsockopen('unix://'.$this->m_strUnix, -1, $nErrorNo, $strError)) !== false)
    {
      $bExit = false;
      fwrite($handle, json_encode($request)."\n");
      while (!$bExit && !feof($handle))
      {
        if (($strLine = fgets($handle)) !== false)
        {
          $bExit = true;
          $response = json_decode($strLine, true);
          if (isset($response['Status']) && $response['Status'] == 'okay')
          {
            $bResult = true;
          }
          else if (isset($response['Error']) && $response['Error'] != '')
          {
            $this->setError($response['Error']);
          }
          else
          {
            $this->setError('Encountered an unknown error.');
          }
        }
      }
      if (!$bResult && $this->getError() == '')
      {
        $this->setError('Failed to receive Warden response.');
      }
      fclose($handle);
    }
    else if ($strError != '')
    {
      $this->setError('fsockopen('.$nErrorNo.') '.$strError);
    }
    else
    {
      $this->setError('Could not connect Warden socket.');
    }

    return $bResult;
  }
  // }}}
  // }}}
  // {{{ setApplication()
  public function setApplication($strApplication)
  {
    $this->m_strApplication = $strApplication;
  }
  // }}}
  // {{{ setError()
  public function setError($strError)
  {
    $this->m_strError = $strError;
  }
  // }}}
  // {{{ vault
  // {{{ vault()
  public function vault($strFunction, $keys, &$data)
  {
    $bResult = false;
    $subData = $data;

    $data = [];
    if ($this->m_strApplication != '')
    {
      $data['Function'] = $strFunction;
      if (!is_array($keys))
      {
        $keys = [];
      }
      array_unshift($keys, $this->m_strApplication);
      $data['Keys'] = $keys;
      if ($subData != null)
      {
        $data['Data'] = $subData;
      }
      $bResult = $this->request('vault', $data);
    }
    else
    {
      $this->setError('Please provide the Application.');
    }
    unset($subData);

    return $bResult;
  }
  // }}}
  // {{{ vaultAdd()
  public function vaultAdd($keys, &$data)
  {
    return $this->vault("add", $keys, $data);
  }
  // }}}
  // {{{ vaultRemove()
  public function vaultRemove($keys)
  {
    return $this->vault("remove", $keys);
  }
  // }}}
  // {{{ vaultRetrieve()
  public function vaultRetrieve($keys, &$data)
  {
    return $this->vault("retrieve", $keys, $data);
  }
  // }}}
  // {{{ vaultRetrieveKeys()
  public function vaultRetrieveKeys($keys, &$data)
  {
    return $this->vault("retrieveKeys", $keys, $data);
  }
  // }}}
  // {{{ vaultUpdate()
  public function vaultUpdate($keys, &$data)
  {
    return $this->vault("update", $keys, $data);
  }
  // }}}
  // }}}
  // {{{ windows
  // {{{ windows()
  public function windows(&$data)
  {
    return $this->request('windows', $data);
  }
  // }}}
  // {{{ windowsUser()
  public function windowsUser($strUser, $strPassword)
  {
    $bResult = false;

    $data = [];
    $data['User'] = $strUser;
    $data['Password'] = $strPassword;
    $bResult = $this->windows($data);
    unset($data);

    return $bResult;
  }
  // }}}
  // }}}
}
?>

File: python/Warden.py
import socket
import os


class Warden:
    # Global arguments across all instances
    m_application = None
    m_unixsock = None
    m_error = None

    def __init__(self, application, unixsock):
        self.m_application = application
        self.m_unixsock = unixsock

    # def __del__(self):
    #    self.m_strConf = None

    def authn(self, data=dict()) -> tuple:
        return self.request("authn", data)

    def authz(self, data=dict()) -> tuple:
        return self.request("authz", data)

    def bridge(self, data: dict) -> tuple:
        return self.request("bridge", data)

    def bridge_authn(self, user: str, pwd: str) -> tuple:
        return self.bridge_authz(user, pwd)

    def bridge_authz(self, user: str, pwd: str, data=dict()) -> tuple:
        data["User"] = user
        data["Password"] = pwd
        return self.bridge(data)

    def central(self, data: dict) -> tuple:
        return self.request("central", data)

    def central_user(self, user: str, data: dict) -> tuple:
        data["User"] = user
        return self.central(data)

    def password(self, data: dict) -> tuple:
        return self.request("password", data)


    def password_application(self, app: str, user: str, pwd: str, stype='') -> tuple:
        request = {"Application": app, "User": user, "Password": pwd}
        if stype:
            request["Type"] = stype
        return self.password(request)


    def password_user(self, user: str, pwd: str) -> tuple:
        request = {"User": user, "Password": pwd}
        return self.password(request)


    def request(self, module: str, request: dict) -> tuple:
        request['Module'] = module;
        (status, response) = self.send(request)
        if status:
            if response and response["Data"]:
                return (status, response["Data"])
            else:
                return (status, dict())
        else:
            return (status, response)


    def send(self, request: dict) -> tuple:
        if os.path.exists(self.m_unixsock):
            with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as sock:
                sock.connect(self.m_unixsock)
                bExit = False
                serialized = json.dumps(request)
                serialized += "\n"
                self.logger.debug("sending: "+serialized)
                sock.sendall(serialized.encode())
                strBuff = ""
                while not bExit:
                    recBytes = sock.recv(4096)
                    if recBytes:
                        strBuff += str(recBytes, 'utf-8')
                        if strBuff.find("\n") != -1:
                            outargs = json.loads(strBuff[0:strBuff.find("\n")])
                            strBuff = strBuff[strBuff.find("\n")+1:]
                            # self.logger.debug("full response: "+json.dumps(outargs))
                            if len(strBuff) <= 0:
                                bExit = True
                            if outargs.get("Status") == "okay":
                                bResult = True
                            if outargs.get("Response"):
                                return (True, outargs["Response"])
                            if outargs.get("Error"):
                                if type(outargs["Error"]) is dict:
                                    self.logger.error(outargs["Error"])
                                    return (False, {"Type": "Bridge", "SubType": "request",
                                                    "Message": outargs["Error"]})
                    else:
                        bExit = True

            return (False, {"Type": "Warden", "SubType": "request",
                            "Message": "found multiple response lines."})
        else:
            return (False, {"Type": "Warden", "SubType": "request",
                            "Message": "Could not connect Warden socket."})


    def set_application(self, application: str) -> None:
        self.m_application = application

    def vault(self, function: str, keys=list(), data=dict()) -> tuple:
        if self.m_application:
            request = dict()
            request["Function"] = function
            keys.insert(0, self.m_application)
            request["Keys"] = keys
            if data:
                request["Data"] = data
            (status, resp) = self.request("vault", request)
            return (status, resp)
        else:
            return (False, {"Type": "Warden", "SubType": "request",
                            "Message": "Please provide the Application."})

    def vault_add(self, keys=list(), data=dict()) -> tuple:
        return self.vault("add", keys, data)

    def vault_remove(self, keys=list()) -> tuple:
        return self.vault("remove", keys)

    def vault_retrieve(self, keys=list()) -> tuple:
        return self.vault("retrieve", keys)

    def vault_retrieve_keys(self, keys=list()) -> tuple:
        return self.vault("retrieveKeys", keys)

    def vault_update(self, keys=list(), data=dict()) -> tuple:
        return self.vault("update", keys, data)
